sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/transformExpressions.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.connector.catalog.functions.BoundFunction
+import org.apache.spark.sql.types.DataType
+
+abstract class TransformExpression extends Expression with Unevaluable {
```

Comment:
I'm still not convinced that this should be an abstract class. I think this doesn't need any subclasses and we can identify any transform function with a `function: BoundFunction` and `children: Seq[Expression]`, including Spark's own bucket function.

Spark's bucket function can have a fake `BoundFunction` which can't be evaluated, but just a marker to indicate it's spark bucket function with a special `canonicalName`.

Reply:
BTW, the numBucket is also part of `children`, as an int literal.

Reply:
> Spark's bucket function can have a fake BoundFunction which can't be evaluated, but just a marker to indicate it's spark bucket function with a special canonicalName.

Hmm how do you make sure the special `canonicalName` is not used by DSv2 implementors? create some reserved name in Spark?

What do you think of keeping a `BucketTransformExpression` with this change? I'm not sure if there is a good way to differentiate the bucket case versus non-bucket case in `TransformExpression` and also check the value of `numBucket` accordingly.

Reply:
> Hmm how do you make sure the special canonicalName is not used by DSv2 implementors?

It's the same problem for other data sources, right? How can they make sure their `canonicialName`s don't conflicts with other data sources.

I think we are fine if the name is special enough: `spark_murmur3_hash_bucket_function`

Reply:
Yea you are right. If necessary, we can also create a special `BoundFunction` sub-class like `SparkBucketFunction` and handle it separately in `TransformExpression.isSameFunction`.